### PR TITLE
feat(client, learn): add helper functions for common validation operations

### LINKED ISSUE #38584: helper functions for challenge tests across curriculum
Implement all the features that was attempted in the https://github.com/freeCodeCamp/freeCodeCamp/pull/38358

### PR DESCRIPTION
**Pull request by hassaanp**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #38584

### CODE REVIEW

**Comment by hassaanp** [comment]
@raisedadead can you look this over quickly - it is mostly the original code by @RandellDawson except for some variable renaming that I thought more appropriate.

I have tested it against one challenge and is working as it should. I will refine this a bit further and add the unit tests.

I presume that the set of functions has to be extended for other use-cases as well.

File: client/src/client/workers/test-evaluator.js
Comment (ojeytonwilliams) [inline client/src/client/workers/test-evaluator.js:61]:
```suggestion
  const __helpers = curriculumHelpers;
```
Any variables we declare here can clash with user defined code.  So, if a user declares `var helpers = {}` then tests relying on `curriculumHelpers` will fail.

While it is not a perfect solution, I think a slightly unusual name like `__helpers` is less likely to be created by users.

**Review by moT01 — COMMENTED** [review]
I gave this a good test and it's looking good to me @hassaanp. Yes, I presume that, as we potentially come up with more helper functions, they will be added in here.

You could possibly make those options configurable. Just a thought, not necessary though.

Also, if you want, you could probably add something about these in the [docs](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/docs/how-to-work-on-coding-challenges.md#writing-tests). Also not necessary, we can make another PR for that if needed.

Edit: actually, I do see a problem. It looks like JS comments don't get removed when they are in a `script` tag. Maybe we should go back to regex.

File: client/src/client/frame-runner.js
Comment (ojeytonwilliams) [inline client/src/client/frame-runner.js:38]:
```suggestion
  const __helpers = curriculumHelpers;
```
This is the inverse of the problem below (for `test-evaluator`).  In this case user defined variables are overwritten by the test code, so if their code relies on them it could fail. 

While this is less likely to be a problem than for the JavaScript challenges, we should name the helpers consistently between the challenges.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/replace-loops-using-recursion.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/replace-loops-using-recursion.english.md:60]:
```suggestion
    testString: assert(!__helpers.removeJSComments(code).match(/for|while|forEach|map|filter|reduce/g));
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/replace-loops-using-recursion.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/replace-loops-using-recursion.english.md:62]:
```suggestion
    testString: assert(__helpers.removeJSComments(sum.toString()).match(/sum\(.*\)/g).length > 1);
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hey @hassaanp, I know this is just a draft, but I saw the change in the naming and it could cause some problems:

File: client/src/client/workers/test-evaluator.js
Comment (moT01) [reply client/src/client/workers/test-evaluator.js:61]: I'm not sure this is true. I thought I gave that a test. But probably not a bad idea to make it more unique anyway. Yea, I could not overwrite or reassign the helpers with a quick test on a JS challenge.

Comment (ojeytonwilliams) [reply client/src/client/workers/test-evaluator.js:61]: It does reassign them - the user's code is `eval`ed after `helpers` is defined.  You can test it like this:

1. Go to http://localhost:8000/learn/javascript-algorithms-and-data-structures/basic-javascript/replace-loops-using-recursion
2. enter the following 
```js
function sum(arr, n) {
  // Only change code below this line
  if(n <= 0) {
    return 0;
  } else {
    return sum(arr, n - 1) + arr[n - 1];
  }
  // Only change code above this line
}

var helpers = {};
```
3. run tests.
4. see `TypeError: helpers.removeJSComments is not a function` in the browser console.

**Comment by ojeytonwilliams** [comment]
>  actually, I do see a problem. It looks like JS comments don't get removed when they are in a script tag. Maybe we should go back to regex.

Good catch! Not sure we need to ditch parsing, though.  We could find the code inside scripts and apply the parser to it.  Alternatively, we could leave a not in the docs to warn people it doesn't apply to html challenges.  There currently aren't any html challenges that rely on `removeJSComments`.

File: client/src/client/frame-runner.js
Comment (hassaanp) [reply client/src/client/frame-runner.js:38]: Agreed.

**Comment by hassaanp** [comment]
> > actually, I do see a problem. It looks like JS comments don't get removed when they are in a script tag. Maybe we should go back to regex.
> 
> Good catch! Not sure we need to ditch parsing, though. We could find the code inside scripts and apply the parser to it. Alternatively, we could leave a not in the docs to warn people it doesn't apply to html challenges. There currently aren't any html challenges that rely on `removeJSComments`.

I'll see what I can do. Maybe we can find a combination of the regex and parsing.

**Comment by hassaanp** [comment]
@ojeytonwilliams what if there are multiple script tags? Or if the script tag has other attributes like `type`? (although that can be accounted for using regex). I think this specific helper (`removeJSComments`) should be used for the JS challenges that don't require a script tag. But if having a utility to find code inside a `script` tag is completely necessary, then I can add a separate utility function that returns the code inside any script tag as an array. Each element in the array will be code inside one script tag.

Let me know if that sounds good @moT01.

**Comment by moT01** [comment]
I know that some of the future projects will have js in a script tag , so we should try to accommodate. And actually, there will be at least one with multiple script tags as well. Both of those are from one of my projects 😕, but I think there will likely be others.

I don't think it would be too tough to come up with some regex to isolate the code in the tags. [Here's one of the files](https://github.com/freeCodeCamp/freeCodeCamp/blob/next-curriculum/curriculum/challenges/english/04-data-visualization/d3-dashboard/part-136.english.md) from the project I made for an example.

**Comment by hassaanp** [comment]
> I know that some of the future projects will have js in a script tag , so we should try to accommodate. And actually, there will be at least one with multiple script tags as well. Both of those are from one of my projects , but I think there will likely be others.
> 
> I don't think it would be too tough to come up with some regex to isolate the code in the tags. [Here's one of the files](https://github.com/freeCodeCamp/freeCodeCamp/blob/next-curriculum/curriculum/challenges/english/04-data-visualization/d3-dashboard/part-136.english.md) from the project I made for an example.

Thanks for the link to the example. I will take a look and adjust the code accordingly.

**Comment by ojeytonwilliams** [comment]
One approach could be to add an extra helper, `removeScriptComments`, that finds all the script elements and then uses `removeJSComments` on them.  That's probably more robust than trying to make `removeJSComments` do everything.

**Comment by hassaanp** [comment]
> One approach could be to add an extra helper, `removeScriptComments`, that finds all the script elements and then uses `removeJSComments` on them. That's probably more robust than trying to make `removeJSComments` do everything.

Yes - that was my perspective as well.

**Comment by hassaanp** [comment]
Added a new function `removeScriptComments` that uses regex to find out any number of `<script>` tags, concatenate the code and remove comments using `removeJSComments` function. I have tested this against different challenges. I suggest the reviewers test it as well.

**Comment by ojeytonwilliams** [comment]
This looks nice, but if I'm reading this right, this discards all the code that's not inside script tags.  @moT01 is that okay, or do you need to get all the code back minus the comments?

**Comment by moT01** [comment]
This is looking good @hassaanp.
 
That looks to be accurate @ojeytonwilliams. It probably wouldn't be a problem and I can't think of or find any examples where this would come into play. But it's not a bad idea to try and keep that code. I would say to give it a try and see if it's not too tough to come up with something. If it's being difficult, maybe we can just go with this. What do you think @hassaanp?

Also, I would try not to replace many more of the functions in the challenges with these to keep this easier to review.

**Comment by moT01** [comment]
On another note. I used a lot of regex to get the styles for a single selector when I was making some projects. Another function we could maybe add is something like that.

```
function getStylesForCssSelector(selector) {
  ...
}
```

or possibly multiple functions...

```
getStylesForIdSelector(s)
getStylesForClassSelector(s)
```

[Here's an example of that](https://github.com/freeCodeCamp/freeCodeCamp/blob/next-curriculum/curriculum/challenges/english/01-responsive-web-design/css-variables-skyline/part-12.md#tests). We don't have to make that if we don't want to right now, but it might be useful.

**Comment by hassaanp** [comment]
> This is looking good @hassaanp.
> 
> That looks to be accurate @ojeytonwilliams. It probably wouldn't be a problem and I can't think of or find any examples where this would come into play. But it's not a bad idea to try and keep that code. I would say to give it a try and see if it's not too tough to come up with something. If it's being difficult, maybe we can just go with this. What do you think @hassaanp?
> 
> Also, I would try not to replace many more of the functions in the challenges with these to keep this easier to review.

Hey, @moT01 yeah that sounds simple enough - what I can do is that I can instead create two functions: one that extracts JS code (possible use-case could be to do JS specific testing), and another that just removes the comments from the `script` tags.

What do you think?

**Comment by hassaanp** [comment]
> On another note. I used a lot of regex to get the styles for a single selector when I was making some projects. Another function we could maybe add is something like that.
> 
> ```
> function getStylesForCssSelector(selector) {
>   
> }
> ```
> 
> so if you were to pass a selector in there, it would return all the styles.
> 
> or possibly multiple functions...
> 
> ```
> getStylesForIdSelector(s)
> getStylesForClassSelector(s)
> ```
> 
> [Here's an example of that](https://github.com/freeCodeCamp/freeCodeCamp/blob/next-curriculum/curriculum/challenges/english/01-responsive-web-design/css-variables-skyline/part-12.md#tests). We don't have to make that if we don't want to right now, but it might be useful.

This makes sense - sure, I will give it a go

**Comment by hassaanp** [comment]
hey guys - sorry for delaying this so long - I have been busy with some personal stuff. My wife and I have had a new baby girl to celebrate, and I just got bogged down with a lot of chores.

I am finally back and will be resuming my contributions.

@ojeytonwilliams @moT01 @raisedadead,  what do you guys think about this PR. Is this still relevant?

Let me know what else you want me to add here.

**Comment by ojeytonwilliams** [comment]
> My wife and I have had a new baby girl to celebrate

:tada: Congratulations!  No need to apologise!

> what do you guys think about this PR. Is this still relevant?

It should be, yes.  My main concern is the one I discussed here: https://github.com/freeCodeCamp/freeCodeCamp/pull/38358#issuecomment-598827498
namely that it could add quite a bit to the size of the site.  I did try to solve that, but as yet I have not come up with a satisfactory fix.  It is something we could return to later, though.

Could you fix the merge conflict?  Then we will be able to test it locally.

**Comment by hassaanp** [comment]
Thanks @ojeytonwilliams !

I have resolved the merge conflicts and rebased the branch.
There were some linting problems in the upstream code that I had to fix in order to pass the pre-commit hooks.

**Comment by ojeytonwilliams** [comment]
This is looking good!  

@hassaanp could you write some unit tests for `curriculumHelpers.js`?  Other than that and a little clean up I think this is good to go.  Ideally the changes to the curriculum should be in a separate commit from the changes to the code and the again from the tests (three commits total).  Also, there's a bit of noise (changes to whitespace) that should probably go.

File: curriculum/challenges/spanish/06-information-security-and-quality-assurance/information-security-and-quality-assurance-projects/anonymous-message-board.spanish.md
Comment (SamuelSHernandez) [inline curriculum/challenges/spanish/06-information-security-and-quality-assurance/information-security-and-quality-assurance-projects/anonymous-message-board.spanish.md:60]: Grammar Error: 

 Cree una aplicación de JavaScript de pila completa que sea funcionalmente similar a esto
**+  "Cree una aplicación de JavaScript de pila completa que sea funcionalmente similar a esta"**

**Review by SamuelSHernandez — CHANGES_REQUESTED** [review]
The Spanish Version has some minimal grammar Issues that need to be corrected.

File: curriculum/challenges/spanish/06-information-security-and-quality-assurance/information-security-and-quality-assurance-projects/anonymous-message-board.spanish.md
Comment (raisedadead) [reply curriculum/challenges/spanish/06-information-security-and-quality-assurance/information-security-and-quality-assurance-projects/anonymous-message-board.spanish.md:60]: Thanks for the review comment @SamuelSHernandez. 

However, the translation is not a part of the changes in this PR per say. If you are interested, you are welcome to make a separate PR :)

I am going ahead and marking this comment resolved.

**Comment by hassaanp** [comment]
@ojeytonwilliams can you please check the test scope in the new file I have pushed `curriculumHelpers.test.js`

I think that the coverage is going to be satisfactory with the planned out tests. Let me know what you think.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
They seem reasonable.  I think the logic is flipped for some, though!

**Comment by hassaanp** [comment]
> They seem reasonable. I think the logic is flipped for some, though!

Yeah, sorry about that. I was not thinking and then I just copied the test over without proof-reading.

Thanks for the quick review!

**Comment by ojeytonwilliams** [comment]
No worries, I assumed that's what you were going for!

**Comment by raisedadead** [comment]
> My wife and I have had a new baby girl to celebrate, and I just got bogged down with a lot of chores.

Congratulations. 🎉 !

**Comment by RandellDawson** [comment]
@ojeytonwilliams I added the comments about the need for tests for the two functions that work with the `script` element.  This PR is now ready for review.  @scissorsneedfoodtoo Can you review also?

If both approve, then I will rebase one last time, so there are only two commits (one for the helper functions (4 files) and one for the 51 challenge files that incorporate the new helper functions.

I have also changed the target branch to `next-curriculum`, since we could use some of these now and @ojeytonwilliams  still wants to figure out a way to implement the Babel differently. 

FYI - I think the one function named `getScriptElementCode` could be rewritten with the following at a later date.  It avoids trying to parse the code with regex.
```js
const getScriptElemCode = () => {
  const previewPane = document.body.querySelector('#fcc-main-frame');
  const doc = previewPane.contentDocument || previewPane.contentWindow.document;
  const scriptElem = [...doc.children].filter(
    child => child.nodeName === 'SCRIPT'
  );
  if (scriptElem.length) {
    return scriptElem[0].toString();
  }
  return '';
};
```

**Comment by RandellDawson** [comment]
@raisedadead I changed the target branch from `master` to `next-curriculum`.  It went from two commits and 55 files to 79 commits and 834 files.  What is the best way to handle this?

**Comment by raisedadead** [comment]
Rebase it against whatever branch you want to target.

**Comment by scissorsneedfoodtoo** [comment]
Tested locally on a few new challenges and everything LGTM.

**Comment by ojeytonwilliams** [comment]
I'll rebase it to next-curriculum in a sec.

File: client/src/utils/curriculum-helpers.js
Comment (ojeytonwilliams) [inline client/src/utils/curriculum-helpers.js:39]:
```suggestion
    let noScripts = codeStr.replace(
      /<script[\s\S]*?>[\s\S]*?<\/script>/gi,
      ''
    );
    codeStr.match(/<script[\s\S]*?>[\s\S]*?<\/script>/gi).forEach(script => {
      const scriptTags = script.match(/(<([^>]+)>)/gi);
      const jsCode = script.replace(/(<([^>]+)>)/gi, '');
      noScripts +=
        scriptTags[0] + removeJSComments(jsCode) + '\n' + scriptTags[1];
    });
    return noScripts;
```
We're not sanitizing here, so we shouldn't imply that we are.

**Review by ojeytonwilliams — DISMISSED** [review]
Just the one nitpick, looks good otherwise!

File: client/src/utils/curriculum-helpers.js
Comment (RandellDawson) [reply client/src/utils/curriculum-helpers.js:39]: Shouldn't the variable name be something more like `noScriptElementComments` instead of `noScripts`?

**Comment by RandellDawson** [comment]
@ojeytonwilliams Upon further review of the two functions (shown below) added by the OP, neither of them do what they are supposed to do, so I have removed them (but this PR will serve as a reminder as to what was attempted).  We were not using them yet and really should be using a different method to accomplish this task (maybe an HTML parser like ACORN).
```js
const removeScriptElementComments = codeStr => {
  try {
    let sanitizedCode = codeStr.replace(
      /<script[\s\S]*?>[\s\S]*?<\/script>/gi,
      ''
    );
    codeStr.match(/<script[\s\S]*?>[\s\S]*?<\/script>/gi).forEach(script => {
      const scriptTags = script.match(/(<([^>]+)>)/gi);
      const jsCode = script.replace(/(<([^>]+)>)/gi, '');
      sanitizedCode +=
        scriptTags[0] + removeJSComments(jsCode) + '\n' + scriptTags[1];
    });
    return sanitizedCode;
  } catch (err) {
    return codeStr;
  }
};

const getScriptElementCode = codeStr => {
  try {
    let concatenatedCodeFromScriptTags = codeStr
      .match(/<script[\s\S]*?>[\s\S]*?<\/script>/gi)
      .reduce((code, scriptTag) => {
        code += scriptTag.replace(/(<([^>]+)>)/gi, '') + '\n';
        return code;
      }, '');
    return removeJSComments(concatenatedCodeFromScriptTags);
  } catch (err) {
    return codeStr;
  }
};
```

The following code string:
```js
const htmlWithScriptElement = `
<html>
  <head>
    <script>
      var text = 'This is text';
      // this comment should be removed
      var test = \`
        should not remove the following script element comment because it is
        part of a text string:
        <script>
          var one = 1;
          /* a comment */
          var two = 2;
        </script>\';
      var three = 3;
      /*
      This multi-line
      comment should be
      removed
      */
      var four = 4;
    </script>
  </head>
  <body>
    <h1>Hello World</h1>
        <script>
      var text2 = 'script inside body element';
      // this comment should be removed
      var test2 = \`
        should not remove the following script element comment because it is
        part of a text string:
        <script>
          var five = 5;
          /* a comment */
          var six = 6;
        </script>\';
      var seven = 7;
      /*
      This multi-line
      comment should be
      removed
      */
      var eight = 8;
    </script>
  </body>
</html>
`;

getScriptElemntCode(htmlWithScriptElement );
```
would result in the script elements in the `test` and `test2` strings being removed, when they are not actual script elements.  This is an example of why a parser should be used to do this instead of regex.

For now to extract the `script` element's code from the editor, we can simply use:
```
const scriptElementCode = $('script')
```

**Comment by RandellDawson** [comment]
@raisedadead This PR seems to be failing due to it just timing out during the build on Travis.  If we close and reopen it, will it use the GitHub Actions testing instead of using Travis?

**Comment by raisedadead** [comment]
Darn - I just relised this will not be able to use the GitHub actions yet, because the `next-curriculum` and the OP branch are not upto date with the `master` branch. We need a rebase on both of the branches to pick-up/remove the CI templates for them to run/ignore.

The simplest way would be to let me know if this looks good and I can merge this as is (abusing my admin rights a bit) as a quick woraround.

Let me know.

**Comment by RandellDawson** [comment]
@raisedadead If @ojeytonwilliams and @ahmadabdolsaheb conclude we can merge next-curriculum to master, then it is a non-issue.

**Comment by raisedadead** [comment]
> ... conclude we can merge next-curriculum to master, then it is a non-issue.

I am not sure that it is related to this status checks per say. If you can give me a go ahead that the changes are good and this has been tested locally, I can make the exception and merge this as is into the `next-curriculum` without the checks passing.

**Comment by RandellDawson** [comment]
I think the changes are fine.  @scissorsneedfoodtoo and @ojeytonwilliams  approved once before.  Maybe @ojeytonwilliams should look at it again?

**Comment by raisedadead** [comment]
Thanks I have updated the battle-plan and assigned @scissorsneedfoodtoo to review this once more.

**Comment by scissorsneedfoodtoo** [comment]
Tested locally with some new curriculum challenges and by running the test suite. Everything LGTM.

**Comment by ojeytonwilliams** [comment]
Sorry for the late comment, but I think the failing tests need to be addressed before we merge this.  Travis timed out because the tests are extremely slow.  That's not necessarily a problem (if the timeout can be increased enough), but if the tests are slow in the browser that impacts user experience.

**Comment by ojeytonwilliams** [comment]
For reference, master:
TestsOnMaster
next-curriculum:
Next-curriculum

and this PR:
Helpers

So, roughly 4 minutes while only 47 challenges are being tested.

**Comment by ojeytonwilliams** [comment]
Did a little more profiling using "test": "mocha --delay --reporter spec --bail".  None of the tests took much more than a few seconds and the specific tests that used `removeJSComments` weren't particularly bad (sub 1 second).  

It seems that all the tests take a little longer rather than specific tests taking a lot longer, which is broadly fine.

As long as we can get the CI to give us enough time, it *should* be okay.

**Comment by RandellDawson** [comment]
@ojeytonwilliams @raisedadead Good news!  Now that we are targeting `master`, this PR no longer has a timeout during the testing phase.

**Comment by raisedadead** [comment]
Thanks for confirming, @moT01 and @scissorsneedfoodtoo if you could give us a sanity check on this, we are ready to land this on master.

**Comment by ojeytonwilliams** [comment]
> @ojeytonwilliams @raisedadead Good news! Now that we are targeting `master`, this PR no longer has a timeout during the testing phase.

Huh.  Was this being run on Travis before and via an action now?

**Comment by raisedadead** [comment]
> Was this being run on Travis before and via an action now?

Yes.

**Pull request merged**
